Doc overhaul

adamgayoso opened the pull request:
This is a version of Sphix-book-based scanpy docs.

**What's lost?**
- Docsearch
- Edit on github button (was weird anyway as no other package does that)

**What's gained?**
- Aesthetics, easier to read
- Git submodule for tutorials
- sphinx.ext.viewcode so each function has a source button (effectively replaces edit on github button)

**What else changed?**
- I removed the release latest stuff, not sure what value it was adding
- API docs are split logically

View the rendered docs here:

https://icb-scanpy--2220.com.readthedocs.build/en/2220/index.html

What's left to do:

- [x] Update tutorials index page to use nbgallery feature of nbsphinx
- [x] Make home index page use a grid of cards, two columns
- [x] Add to contributing guide how to manage the git submodule (really just occasionally do `git submodule update --remote`)
- [x] Annotate CSS overrides
- [x] Exchange sphinx ext viewcode for linkcode (probably in scanpydoc)

adamgayoso committed:
Update conf.py

adamgayoso committed:
add req

adamgayoso committed:
add override path

adamgayoso committed:
Create override.css

adamgayoso committed:
add pygments, remove scvi-tools colors

adamgayoso committed:
add myst extensions

ivirshup commented on docs/conf.py:
```diff
@@ -79,11 +78,19 @@
 todo_include_todos = False
 api_dir = HERE / 'api'  # function_images
+myst_enable_extensions = [
+    "colon_fence",
```
```suggestion
```

Do we need this?

adamgayoso replied:
I mean it's nice to have, especially once I add sphinx design elements

ivirshup replied:
What does it do?

My impression was that it was just for compatibility with [some markdown editor](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#code-fences-using-colons).

adamgayoso replied:
if I understand correctly, it would make the docs easier to view in a standard markdown viewer like github of vscode that's not doing doc rendering

ivirshup replied:
Can you show me the docs for this? When I did a cursory search before I only found one markdown editor that uses these.

adamgayoso replied:
It's recommended by [sphinx design](https://sphinx-design.readthedocs.io/en/furo-theme/get_started.html)

adamgayoso committed:
remove github edit links

adamgayoso committed:
move css

adamgayoso committed:
add sphinx design, opengraph

adamgayoso committed:
move contributors out of index

adamgayoso committed:
show source

adamgayoso committed:
add submodule

adamgayoso committed:
add submodule

adamgayoso committed:
temp old tutorials rename

adamgayoso committed:
remove github links

adamgayoso committed:
remove github url from template

adamgayoso committed:
disable warning fail

adamgayoso committed:
disable warning fail

adamgayoso committed:
nbsphinx

adamgayoso committed:
fix headings

adamgayoso committed:
rtd use submodule

adamgayoso committed:
add viewcode

adamgayoso committed:
more headings fixes

adamgayoso committed:
tutorials index

adamgayoso committed:
add nbsphinx ext

adamgayoso committed:
remove sphinx pygments

adamgayoso committed:
fix colab link

adamgayoso committed:
add ipython as docs req

adamgayoso committed:
no signatures in tables

adamgayoso committed:
make ci happy

flying-sheep reviewed (changes requested):
Looks great, and I like the submodule approch for tutorials.

- [ ] One thing I’m missing is thorough documentation of the CSS overrides.
  I noticed that when things get updated, it’s often no longer clear why which rule is there.
- [x] The only specific CSS change I’d like to see is dimming the nbsphinx code cell borders in dark mode, they’re too bright. I don’t see why they need to have borders at all actually …

flying-sheep committed:
Merge branch 'master' into furo

flying-sheep commented on docs/_templates/autosummary/class.rst:
```diff
@@ -1,5 +1,3 @@
-:github_url: {{ fullname | github_url }}
```
Hmm, this is a regression in my eyes. The `[source]` links now link to `_modules/` URLs. I preferred them linking to GitHub. Can this be done?

adamgayoso replied:
maybe this?

https://www.sphinx-doc.org/en/master/usage/extensions/linkcode.html

flying-sheep replied:
That looks good!

adamgayoso replied:
it's been added

ivirshup commented:
Can we customize the contents of left hand table of contents drop down?

ivirshup commented on docs/conf.py:
```diff
@@ -54,16 +53,18 @@
     'scanpydoc.definition_list_typed_field',
     'scanpydoc.autosummary_generate_imported',
+    "sphinx_design",
+    "sphinxext.opengraph",
```
What does this give us?

adamgayoso replied:
sphinx design -- nothing yet, but would be nice to have some cards included from there

open graph -- this gives link preview when you share a doc link online (e.g., twitter)

ivirshup reviewed (commented):
> Docsearch

Gotta keep this (it's so much nicer)

But this is part of the reason the styling parts should end up in scanpydoc, right?

----------------

<details>
<summary> LHS TOC gets cut off too easily </summary>

This isn't great:

![image](https://user-images.githubusercontent.com/8238804/161974421-168d4c5c-8a95-4557-bd7a-e40fdbd32b44.png)


Especially when there is clearly space for it:

![image](https://user-images.githubusercontent.com/8238804/161974599-8e20525e-a7fb-4f1e-a460-4ee234cdb4a1.png)

</details>


Function names should also be monospaced

---------------

This broke github links. I do think can pretty easily fix that with `sphinx.ext.extlinks` though ([example](https://github.com/pydata/xarray/blob/facafac359c39c3e940391a3829869b4a3df5d70/doc/conf.py#L89-L92)).

---------------

I would want some developer docs for handling the tutorials as a gitsubmodule.

My very cursory understanding is the submodule is included at a particular commit, and there are extra steps to cloning them in?

flying-sheep commented:
> > Docsearch
> 
> Gotta keep this (it's so much nicer)

The UX is for sure, but if we could replace it with something that has similar UX but uses an index that’s built by the RTD build, I’d much prefer that. It’s silly that searching in PR builds or `stable` will result in getting redirected to the `latest` docs …

adamgayoso committed:
add sphinx design cards

adamgayoso committed:
break up api

adamgayoso committed:
fix icons

adamgayoso commented:
@ivirshup @flying-sheep I just added some sphinx design cards to the main page and split out the API docs into multiple pages, which looks nicer in the left sidebar. If you guys like it we can do the same for the external API

adamgayoso committed:
link to new api index in card

adamgayoso committed:
update colors

adamgayoso committed:
no latest release in index

adamgayoso committed:
release notes

adamgayoso committed:
colors

adamgayoso committed:
color typo

adamgayoso committed:
back to default colors

flying-sheep committed:
Update override.css

flying-sheep commented:
> If you guys like it

I do like it, but it’s a bit spaceous.

- [x] Maybe we can do a 2-wide grid for those cards?

Also I added 268899df6a5d121960fb48970f5553865b5ba967 for more subtle cell borders in dark mode

adamgayoso committed:
add module

adamgayoso committed:
move external

flying-sheep committed:
Make border color override work

adamgayoso commented:
@flying-sheep I put a todo list at the top of the PR. What do you think? Would you be able to help move stuff to scanpydoc?

adamgayoso committed:
Merge branch 'master' into furo

adamgayoso committed:
grid

adamgayoso committed:
hoverxref and linkcode

adamgayoso committed:
fix grid

adamgayoso committed:
comment

adamgayoso committed:
grid spacing

adamgayoso committed:
intersphinx hoverxref

adamgayoso committed:
extlinks

adamgayoso committed:
extlinks

adamgayoso committed:
fix contributors

adamgayoso committed:
fix tutorials to use old page

adamgayoso committed:
hidden toc tree

adamgayoso committed:
toctree for tutorials hidden

adamgayoso committed:
clean up conf

adamgayoso commented:
I've addressed and fixed just about everything @ivirshup @flying-sheep can we get this in?

A future PR should rename scanpydoc to scversedoc and consolidate all these custom additions.

adamgayoso committed:
fix extlinks

adamgayoso commented:
hoverxref/code source links will not work until this gets merged because the 1) the pr build docs are not hosted on the readthedocs domain and two for codelinks the pr build adds a weird nonexistent tag. But I confirmed this all works for scvi-tools

adamgayoso committed:
tutorial extlink

adamgayoso committed:
add leiden to docs build

adamgayoso commented:
It's currently failing now due to it trying to render tutorials in the tutorials repo that aren't actually used (pbmc5k) but the link at in the first comment still shows the last successful build.

adamgayoso committed:
buttons

adamgayoso committed:
small role

adamgayoso committed:
try rst badge

adamgayoso committed:
extlinks

adamgayoso committed:
fix image link

Zethson commented:
@adamgayoso is there any interest in picking this up again? You also mentioned at some point that you might ditch furo for the book theme?

Would be good if we didn't jump from theme to theme :)

adamgayoso commented:
@Zethson this is pretty much good to go and follows almost exactly how we have it in scvi tools

adamgayoso committed:
Merge branch 'master' into furo

Zethson commented on docs/conf.py:
```diff
@@ -107,28 +126,70 @@
+html_title = "scanpy"
+
+
+# def setup(app):
```
We can kill this I guess?

adamgayoso replied:
I commented that out due to the unused tutorials from the other PR. I should add it back; however, we don't use it in cookiecutter

Zethson commented on docs/contributors.md:
```diff
@@ -1,13 +1,22 @@
+- [Giovanni Palla](https://twitter.com/g_palla1), developer, spatial data
+- [Malte Luecken](https://twitter.com/MDLuecken), developer, community & forum
+- [Lukas Heumos](https://twitter.com/LukasHeumos), developer, diverse contributions
+- [Philipp Angerer](https://github.com/flying-sheep), developer, software quality, initial anndata conception ☀
```
@adamgayoso I wouldn't mind it if you added yourself.

Zethson reviewed (changes requested):
```
⋊> ~/P/s/docs on furo  make html                                                                                                                                            (scanpy) 17:28:53
Running Sphinx v5.3.0

Exception occurred:
  File "/home/zeth/miniconda3/envs/scanpy/lib/python3.8/site-packages/scanpydoc/__init__.py", line 12, in <module>
    __version__ = get_version(__file__, vcs="git")
TypeError: get_version() got an unexpected keyword argument 'vcs'
The full traceback has been saved in /tmp/sphinx-err-kj7xpoly.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:25: html] Error 2
```

Can't build the docs yet

adamgayoso commented:
@Zethson docs are built in the CI rtd, not sure why you can't build it locally

Zethson commented:
I killed my scanpy development environment and tried again. Works now smh

Zethson reviewed (approved):
LGTM

Zethson commented:
@adamgayoso tell me when you want it merged, please.

adamgayoso commented on docs/conf.py:
```diff
@@ -107,28 +126,70 @@
+
+
+# def setup(app):
+#     app.warningiserror = on_rtd
```
```suggestion
def setup(app):
    app.warningiserror = on_rtd
```

adamgayoso committed:
Update docs/conf.py

adamgayoso committed:
update submodule

adamgayoso committed:
remove hoverxref

adamgayoso committed:
warning code

adamgayoso committed:
add how to update submodule

adamgayoso committed:
remove unnecessary css

adamgayoso committed:
remove more scanpydoc

adamgayoso commented:
Ok @Zethson it's ready to merge

adamgayoso committed:
tutorials link

Zethson commented:
Thank you @adamgayoso !
We should consider removing the ecosystem page in favor our scverse ecosystem page in the future to not have to maintain both.

adamgayoso committed:
Merge branch 'master' into furo

adamgayoso committed:
switch to sphinx book

adamgayoso committed:
pre commit'

adamgayoso committed:
update reqs

adamgayoso committed:
logo only

adamgayoso committed:
deps

adamgayoso committed:
linkcode

adamgayoso committed:
linkcode

adamgayoso committed:
exclude patterns

adamgayoso committed:
sidebar levels

adamgayoso committed:
remove external tools

adamgayoso committed:
fix external

adamgayoso committed:
Merge branch 'master' into furo

adamgayoso committed:
Merge branch 'master' into furo

flying-sheep commented on docs/conf.py:
```diff
@@ -107,28 +129,37 @@
-    github_version='master',  # Version
-    conf_py_path='/docs/',  # Path in the checkout to the docs root
-)
+html_theme = "sphinx_book_theme"
+html_theme_options = {
+    "repository_url": repository_url,
+    "use_repository_button": True,
+    "logo_only": True,
+    "show_toc_level": 4,  # show all levels in the sidebar
+}
 html_static_path = ['_static']
+html_css_files = ["css/override.css"]
 html_show_sphinx = False
 html_logo = '_static/img/Scanpy_Logo_BrightFG.svg'
+html_title = "scanpy"
 
 
 def setup(app):
-    app.warningiserror = on_rtd
+    """App setup hook."""
+    # TODO: fix all warnings in a future PR
+    # Many come from the tutorials, like the workshop directory
+    # which is not included in the docs
+    #     app.warningiserror = on_rtd
```
Wait a second, how could I miss this.

This is an absolute no-go. This makes it very hard to go back to functional docs where nothing is broken, since broken stuff can accumulate after this.

Someone needs to spend far too much time to fix this, and I fear it’s going to be me. Can we please never do this again?
